### PR TITLE
Feature/latex prep job

### PR DESCRIPTION
**Pull request by pdiercks**
Hi @dglaeser ,
Please have a look at how I handled the tex related tasks. I directly changed this in `dodo.py` to test locally.
I am using a python-action `write_source` in `task_write_paper_source` to write `paper.tex`. 
This python-action should probably be a script to be also used in this way in the other workflow implementations, but I first wanted to know if you approve of the solution or not?

### CODE REVIEW

File: simple_use_case/pydoit/dodo.py [hunk 1]
```diff
@@ -53,28 +55,79 @@ def task_plot_over_line():
...
-    source_code = SOURCE / "paper.tex"
-    root_code = ROOT / "paper.tex"
+def task_write_paper_source():
+    """write paper source to file"""
+
+    def write_source(gmsh_stdout, dependencies, targets):
+        # file names in correct order
+        fnames = ["lineplot.tex", "plotoverline.csv", "numdofs.txt"]
+        deps = fix_deps(dependencies, fnames)
+        preamble = """\\documentclass[12pt]{article}
```
Comment (joergfunger) [inline simple_use_case/pydoit/dodo.py:71]: Instead of writing the complete paper to a file (which does not allow to compile this), I would rather "\input" or "\include" only the macro definition that is then written in this task". This way, someone just interested in compiling the paper manually (while changing the "raw" text in his/her text editor can do so (otherwise you would always have to call the dodo task write_paper_source). In addition, this allows you to work in a standard latex editor that understands the syntax, ..

**Review by joergfunger — APPROVED** [review]
Thanks, just a tiny modification to add the paper.tex as a separate file and input only the macrodefinition.

File: simple_use_case/pydoit/dodo.py (hunk 1 above)
Comment (pdiercks) [reply simple_use_case/pydoit/dodo.py]: so I changed this in the python-action as desired.

File: simple_use_case/pydoit/dodo.py [hunk 2]
```diff
@@ -53,28 +61,67 @@ def task_plot_over_line():
...
+
+        # read file written by task_poisson to determine number of DoFs
+        with open(deps[3], "r") as handle:
+            ndofs = handle.read()
+
+        plot = "\\inputplot{{{}}}{{{}}}{{{}}}{{{}}}\n".format(deps[1], deps[2], float(size), int(ndofs))
+        paper = paper.replace("% TODO inputplot", plot)
+
+        with open(targets[0], "w") as outfile:
+            outfile.write(paper)
```
Comment (pdiercks) [inline simple_use_case/pydoit/dodo.py:93]: While the use of the python-action for this task is natural in `doit`, what bothers me is that using small snippets we end up with the same code snippet (in whatever programming language the tool supports) for each tool (code duplication). We could avoid this by putting the contents of `write_source` in a separate script, but this feels a bit like overkill (with a proper CLI for all the arguments). Also it makes the definition of the `task_write_paper_source` unnecesarry complicated (in the case of `doit`).
Which option do you prefer @joergfunger and @dglaeser ?

**Comment by pdiercks** [comment]
Hi @dglaeser , I made the changes as we discussed earlier. Please have a look and let me know if I missed something.

File: simple_use_case/source/lineplot.tex [hunk 3]
```diff
@@ -0,0 +1,10 @@
+\begin{figure}[h]
+    \centering
+    \begin{tikzpicture}
+        \begin{axis}[]
+            \addplot[color=red,smooth,thick] table[col sep=comma,header=true,x index=2,y index=0] {\DATA};
+        \end{axis}
+    \end{tikzpicture}
+    \caption{Solution of the poisson equation over the line $L = \{(x, y)\in\Omega = {(0, \SIZE{})}^2 | x = y\}$ using \NUMDOFS{} DoFs in the finite element model.}%
+    \label{fig:u}
+\end{figure}
```
Comment (dglaeser) [inline simple_use_case/source/lineplot.tex:10]: shouldn't this file be obsolete now?

File: simple_use_case/source/poisson.py [hunk 4]
```diff
@@ -59,28 +60,39 @@ def solve_and_write_output(mesh: str, degree: int, outputfile: str):
...
     """
     discrete_solution = solve_poisson(mesh, degree)
     discrete_solution.rename("u", discrete_solution.name())
     resultFile = df.File(outputfile)
     resultFile << discrete_solution
 
+    with open(numdofs, "w") as handle:
+        dofs = discrete_solution.function_space().dim()
+        handle.write("{}\n".format(dofs))
+
```
Comment (dglaeser) [inline simple_use_case/source/poisson.py:74]:
```suggestion
    dofs = discrete_solution.function_space().dim()
    with open(numdofs, "w") as handle:
        handle.write("{}\n".format(dofs))
    print(f"Number of dofs used: {dofs}")
```

This way, tools that can parse values without the detour over files (e.g. cwl by grepping stdout) don't have to, right?

File: simple_use_case/source/postprocessing.py [hunk 5]
```diff
@@ -13,40 +12,42 @@
...
+    parser.add_argument("-i", "--inputfile", required=True, help="The source pvd filepath.")
+    parser.add_argument("-o", "--outputfile", required=True, help="The target csv filepath.")
     parser.add_argument(
         "--field",
         type=str,
         default="u",
         help="Field variable to plot (default: u)",
     )
-    args = parser.parse_args(sys.argv[1:])
+    args = vars(parser.parse_args())
```
Comment (dglaeser) [inline simple_use_case/source/postprocessing.py:52]: Is the change to this file related to this PR? It seems to me to just be some cleanup related to runtime arg handling, no? If so, I would do this in a separate PR, because this changes the script usage and thus we have to port all implementations. I guess it's easier to review if that's a separate PR. But I may be mistaking and it is related to the issue adressed in this PR

File: simple_use_case/source/lineplot.tex (hunk 3 above)
Comment (pdiercks) [reply simple_use_case/source/lineplot.tex:10]: yes, I just forgot to remove it.

File: simple_use_case/source/poisson.py (hunk 4 above)
Comment (pdiercks) [reply simple_use_case/source/poisson.py:74]: Yes. `doit` can parse values without the detour over files by the way. But I would keep the additional text file output if one of the tools might not support grepping stdout directly.

File: simple_use_case/source/postprocessing.py (hunk 5 above)
Comment (pdiercks) [reply simple_use_case/source/postprocessing.py:52]: The reason why I made changes to this file is that I had to allow for the arbitrary domain size. Before the line plot was hardcoded to the unit square. IMO this particular commit should be included in #32 which is what this PR is proposing.

File: simple_use_case/source/poisson.py (hunk 4 above)
Comment (dglaeser) [reply simple_use_case/source/poisson.py]: yeah, I agree. That's why I left it in the suggestion.

File: simple_use_case/source/postprocessing.py (hunk 5 above)
Comment (dglaeser) [reply simple_use_case/source/postprocessing.py:52]: yes I saw that you now get the bounds from the vtk file, that's nice. And I agree that that may be better placed in #32, although I have no strong opinion about that because we merge this one into #32, anyway...

But I was talking about your changes to the script arguments. You changed `pvd` and `csv` to `-i` and `-o`, no? I like that, but I think it is a different concern and may be easier to change in a separate PR because we have to adjust the calls to this script in all WF realizations, and these changes would clutter the diff to review here and are also unrelated to the task of this PR. Or maybe I am missing something.. What do you think?

Comment (pdiercks) [reply simple_use_case/source/postprocessing.py:52]: okay, I think you are right. The change for that specific line with the script arguments is a different concern. I just thought it would not make a big difference since we have to adjust all the other workflow implementations to also use domain size and numdofs. 
I have no strong opinion about in which PR this change should be made, so I will just follow your suggestion and revert the change in the CLI for this PR.

**Comment by pdiercks** [comment]
Hi @dglaeser , I would like to merge this now into #32 and then work on the other WF realizations, also to be merged into #32, and the finally merge #32 into main to test everything.

**Comment by dglaeser** [comment]
@pdiercks, thanks a lot for the latest changes, I'll merge this now and then we can continue on #32. I will rebase #34 and get the CWL implementation running with your new changes tomorrow.
